src/persist-client/src/internal/compact.rs

```diff
@@ -184,82 +192,416 @@ impl Compactor {
+            if let Some(part) = parts.next() {
+                let mut part =
+                    fetch_batch_part(&shard_id, blob.as_ref(), &metrics, &part.key, part_desc)
+                        .await?;
+                while let Some((k, v, mut t, d)) = part.next() {
+                    t.advance_by(desc.since().borrow());
+                    let d = D::decode(d);
+                    let k = k.to_vec();
+                    let v = v.to_vec();
+                    // default heap ordering is descending
+                    sorted_updates.push(Reverse((((k, v), t, d), index)));
+                    remaining_updates_by_run[index] += 1;
+                }
+            }
+        }
+
+        // repeatedly pull off the least element from our heap, refilling from the originating run
+        // if needed. the heap will be exhausted only when all parts from all input runs have been
+        // consumed.
+        while let Some(Reverse((((k, v), t, d), index))) = sorted_updates.pop() {
+            remaining_updates_by_run[index] -= 1;
+            if remaining_updates_by_run[index] == 0 {
+                // repopulate from the originating run, if any parts remain
+                let (part_desc, parts) = &mut runs[index];
+                if let Some(part) = parts.next() {
+                    let mut part =
+                        fetch_batch_part(&shard_id, blob.as_ref(), &metrics, &part.key, part_desc)
+                            .await?;
+                    while let Some((k, v, mut t, d)) = part.next() {
+                        t.advance_by(desc.since().borrow());
                         let d = D::decode(d);
-                        updates.push(((k.to_vec(), v.to_vec()), t, d));
+                        let k = k.to_vec();
+                        let v = v.to_vec();
+                        // default heap ordering is descending
+                        sorted_updates.push(Reverse((((k, v), t, d), index)));
+                        remaining_updates_by_run[index] += 1;
                     }
                 }
             }
-            consolidate_updates(&mut updates);
 
-            let len = updates.len();
-            let mut builder = ColumnarRecordsVecBuilder::new_with_len(cfg.blob_target_size);
-            for ((k, v), t, d) in updates.iter() {
-                builder.push(((k, v), T::encode(t), D::encode(d)));
+            let update_size_bytes = ColumnarRecordsBuilder::columnar_record_size(&k, &v);
 
-                // Flush out filled parts as we go to keep bounded memory use.
-                for chunk in builder.take_filled() {
-                    parts
-                        .write(chunk, req.desc.upper().clone(), req.desc.since().clone())
-                        .await;
-                }
+            // flush the buffer if adding this latest update would cause it to exceed our target size
+            if update_size_bytes + update_buffer_size_bytes > cfg.blob_target_size {
```

Comment:
we also need this logic in BatchBuilder, right? let's put a TODO there so we don't forget

Reply:
to mark where we need to make changes for https://github.com/MaterializeInc/materialize/issues/14579?

Reply:
yeah, exactly